extension/storage/filestorage/client.go

```diff
@@ -108,7 +120,20 @@ func (c *fileStorageClient) Batch(_ context.Context, ops ...storage.Operation) e
 		return nil
 	}
 
-	return c.db.Update(batch)
+	c.compactionMutex.RLock()
+	defer c.compactionMutex.RUnlock()
+	err := c.db.Update(batch)
+	if c.shouldCompactOnWrite() {
```

Comment:
This condition can be triggered concurrently. While this doesn't cause any correctness problems, we should try to avoid it. Simplest solution is probably checking this condition again in `Compact()` after taking the lock, but it's awkward with compaction on start, and a bit implicit. A more idiomatic solution would be to have a loop running off a ticker and compacting if the condition is met. Something like:
```go
func (c *fileStorageClient) compactionLoop() {
	ctx := context.Background()
	for {
		select {
		case <-c.stopCh:
			return
		case <-c.compactionTicker.C:
			if c.shouldCompact() {
				err := c.Compact(ctx, c.compactionCfg.Directory, c.openTimeout, c.compactionCfg.MaxTransactionSize)
				if err != nil {
					c.logger.Error("compaction failure",
						zap.String(directoryKey, c.compactionCfg.Directory),
						zap.Error(err))
				}
			}
		}
	}
}
```